### PR TITLE
Channels implementation with new listeners

### PR DESCRIPTION
**Pull request by natsukagami**
Depends on #21 and #22. 

## Channel stuff

- **New implementation for sync and bounded channels**, based on `lockBoth`.
- Remove suspendSupport requirement from channel.
- **New channel kind**: unbounded channels, with `.sendImmediately` non-suspending.
- ChannelMultiplexer changes a bit: there is now a blocking/suspending `start()` method instead of spawning a background working.

## Stuff based on channels

- **Future Collectors**: `Seq[Future[T]]` => `ReadableChannel[Future[T]]`, with values arriving as they are completed.
- Implement collector-based future collection extensions: `Seq[Future[T]]` has `awaitAll`(`OrCancel`) and `altAll`(`WithCancel`)

## Other misc goodies

- Add shorthand `.await` syntax to sources.
- Add `Source.values` to quickly create a new source based on a list of values.

### CODE REVIEW

File: shared/src/main/scala/async/Async.scala [hunk 1]
```diff
@@ -127,6 +131,30 @@ object Async:
...
   end OriginalSource
 
+  object Source:
+    /** Create a [[Source]] containing the given values, resolved once for each. */
+    def values[T](values: T*) =
+      import scala.collection.JavaConverters._
+      val q = java.util.concurrent.ConcurrentLinkedQueue[T](values.asJavaCollection)
+      new Source[T]:
+        override def poll(k: Listener[T]): Boolean =
+          if q.isEmpty() then return false
```
Comment (diesalbla) [inline shared/src/main/scala/async/Async.scala:141]: I am curious: does the `return` give some running time improvement?

Comment (natsukagami) [reply shared/src/main/scala/async/Async.scala:141]: Nice catch! In fact no, it doesnt and may slow it down :P I'll remove it, thanks!

File: shared/src/main/scala/async/channels.scala [hunk 2]
```diff
@@ -272,39 +323,43 @@ object ChannelMultiplexer:
...
+          publishersCopy = publishers.toSeq
+
+        Async.select(
+          (infoChannel.readSource ~~> {
+            case Left(_) | Right(Message.Quit) =>
+              ChannelMultiplexer.this.synchronized:
+                subscribersCopy = subscribers.toList
+              for (s <- subscribersCopy) s.send(Failure(ChannelClosedException()))
+              shouldTerminate = true
+            case Right(Message.Refresh) => ()
```
Comment (m8nmueller) [inline shared/src/main/scala/async/channels.scala:352]: Why does this exist? And if it is only quit left, then I guess the info channel is not needed anymore. You can implement the quit using a (classical) Promise instead.

**Review by m8nmueller — COMMENTED** [review]
The abstraction of channels is great!!

I also changed some things in place, have a look.

File: shared/src/main/scala/async/channels.scala (hunk 2 above)
Comment (natsukagami) [reply shared/src/main/scala/async/channels.scala:352]: `Refresh` is needed as like a `continue` for the while loop. Happens when we made a change to the subscriber/publisher list.

File: shared/src/main/scala/async/channels.scala [hunk 3]
```diff
@@ -6,246 +6,295 @@ import scala.util.{Failure, Success, Try}
...
-            flushListeners()
-          }
       }
-
-    def close(): Unit =
-      BufferedChannel.this.synchronized:
+  }
+end UnboundedChannel
+
+object Channel:
```
[the remark this answers is not in the document]
Comment (natsukagami) [reply shared/src/main/scala/async/channels.scala:168]: I think it is possible (in fact it should be possible to implement completely lock-free channels) but not right now

**Review by m8nmueller — APPROVED** [review]
Nicely done with map. LGTM.

**Pull request merged**